lib/Dialect/TritonGPU/Transforms/DotSlicing.cpp

```diff
@@ -0,0 +1,466 @@
+    // properly. Do it manually.
+    getBackwardSlice(operand, &bwdSlices);
+    std::copy_if(bwdSlices.begin(), bwdSlices.end(),
+                 std::back_inserter(filteredSlices),
+                 [](Operation *op) { return isa<tt::DotOp>(op); });
+
+    if (filteredSlices.empty()) {
+      return false;
+    }
+    return true;
+  }
+
+  bool shouldSliceDot(tt::DotOp dotOp) {
+    auto dotOperand = dotOp.getOperand(0);
+    auto dotATy = dotOperand.getType().cast<RankedTensorType>();
+    auto kDim = dotATy.getShape()[1];
+
+    if (this->sliceKTile == 0 || this->sliceKTile == kDim) {
+      return false;
+    }
+    return true;
+  }
+
+  void dotSlicingDCE(ArrayRef<Operation *> eraseOps) {
+    for (Operation *opToErase : llvm::reverse(eraseOps)) {
+      assert(opToErase);
+      bool hasUses = false;
+      for (auto result : opToErase->getResults()) {
+        if (!result.use_empty()) {
+          hasUses = true;
+        }
+      }
+      if (hasUses) {
+        continue;
+      }
+      opToErase->erase();
+    }
+  }
+
+  Operation *getFirstOpToSlice(tt::DotOp dotOp, int operandIdx) {
+    if (dependsOnPreviousDot(dotOp, operandIdx)) {
+      return dotOp.getOperand(operandIdx).getDefiningOp();
+    }
+    return getLoadInst(dotOp, operandIdx);
+  }
+
+  void runOnOperation() override {
+    getOperation()->walk([&](tt::DotOp dotOp) {
+      if (!shouldSliceDot(dotOp)) {
+        return;
+      }
+
+      OpBuilder builder(dotOp);
+      SmallVector<Operation *> eraseOps;
+
+      auto dotResTy = dotOp.getType().cast<RankedTensorType>();
+      auto dotOperand = dotOp.getOperand(0);
+      auto dotATy = dotOperand.getType().cast<RankedTensorType>();
+      auto dotAShape = dotATy.getShape();
+      int64_t numSlices = dotAShape[1] / this->sliceKTile;
+      Value slicedAcc = dotOp.getOperand(2);
+
+      auto firstOpToSliceA = getFirstOpToSlice(dotOp, 0);
+      auto firstOpToSliceB = getFirstOpToSlice(dotOp, 1);
+
+      if (setLayoutForSlicing(firstOpToSliceA, /*operandIdx*/ 0)) {
+        firstOpToSliceA = getFirstOpToSlice(dotOp, /*operandIdx*/ 0);
+      }
+
+      if (setLayoutForSlicing(firstOpToSliceB, /*operandIdx*/ 1)) {
+        firstOpToSliceB = getFirstOpToSlice(dotOp, /*operandIdx*/ 1);
+      }
+
+      for (int i = 0; i < numSlices; i++) {
+        auto slicedOperandA = getSlicedDotOperand(
+            firstOpToSliceA, dotOp, 0, i, this->sliceKTile, builder, eraseOps);
+        auto slicedOperandB = getSlicedDotOperand(
+            firstOpToSliceB, dotOp, 1, i, this->sliceKTile, builder, eraseOps);
+
+        auto slicedDot = builder.create<tt::DotOp>(
+            dotOp.getLoc(), dotResTy, slicedOperandA, slicedOperandB, slicedAcc,
+            dotOp.getAllowTF32(), dotOp.getMaxNumImpreciseAcc());
+        slicedAcc = slicedDot;
+      }
+
+      eraseOps.push_back((Operation *)dotOp);
+      dotOp.replaceAllUsesWith(slicedAcc);
+      dotSlicingDCE(eraseOps);
```

Comment:
I'm wondering why we still need to keep the dotOp after its uses have been replaced with the slicedAcc?

Reply:
I found that in dotSlicingDCE, the op is only removed if it does not have any uses. So I guess my questions is
- If the old dotOp's uses have been replaced, then why we need to check if it still has uses?

Reply:
For dotOp it's not really needed. However, some instructions we are slicing along the way (the one that are pushed to eraseOps vector) should not be removed for every dot. 

For example, in FA fwd pass, when causal=True, one load is used by 2 dotOps (main loop is sliced into 2 for loops, so everything is doubled). So when we are processing first of these two dot ops, we will slice that load, but still want to keep the original "large" load because it's used by second dot (in "second" for loop). Only when we are processing second dot can we erase the original large load and only use sliced loads.  

Same goes for other instructions with multiple uses that we are slicing along the way. This is why we need dotSlicingDCE. For dotOp however, it will be removed immediately by dotSlicingDCE (we don't expect to have any external uses).